consensus/signatures.go

```diff
@@ -12,220 +12,181 @@ var (
-	return
+	return true
 }
 
 // validCoveredFields makes sure that all covered fields objects in the
-// signatures follow the rules. This means that if `WholeTransaction` is set to
-// true, all fields except for `Signatures` must be empty. All fields must be
+// signatures follow the rules. This means that if 'WholeTransaction' is set to
+// true, all fields except for 'Signatures' must be empty. All fields must be
 // sorted numerically, and there can be no repeats.
-func (t Transaction) validCoveredFields() (err error) {
+func (t Transaction) validCoveredFields() error {
 	for _, sig := range t.Signatures {
-		// Check that all fields are empty if `WholeTransaction` is set.
+		// convenience variables
 		cf := sig.CoveredFields
+		fields := [][]uint64{
+			cf.SiacoinInputs,
+			cf.MinerFees,
+			cf.FileContracts,
+			cf.FileContractTerminations,
+			cf.StorageProofs,
+			cf.SiafundInputs,
+			cf.SiafundOutputs,
+			cf.ArbitraryData,
+		}
+
+		// Check that all fields are empty if 'WholeTransaction' is set.
 		if cf.WholeTransaction {
-			if len(cf.SiacoinInputs) != 0 ||
-				len(cf.MinerFees) != 0 ||
-				len(cf.SiacoinOutputs) != 0 ||
-				len(cf.FileContracts) != 0 ||
-				len(cf.FileContractTerminations) != 0 ||
-				len(cf.StorageProofs) != 0 ||
-				len(cf.SiafundInputs) != 0 ||
-				len(cf.SiafundOutputs) != 0 ||
-				len(cf.ArbitraryData) != 0 {
-				err = errors.New("whole transaction flag is set but not all fields besides signatures are empty")
-				return
+			for _, field := range fields {
+				if len(field) != 0 {
+					return errors.New("whole transaction flag is set, but not all fields besides signatures are empty")
+				}
 			}
 		}
 
 		// Check that all fields are sorted, and without repeat values, and
 		// that all elements point to objects that exists within the
 		// transaction.
-		err = sortedUnique(cf.SiacoinInputs, len(cf.SiacoinInputs)-1)
-		if err != nil {
-			return
-		}
-		err = sortedUnique(cf.MinerFees, len(cf.MinerFees)-1)
-		if err != nil {
-			return
-		}
-		err = sortedUnique(cf.SiacoinOutputs, len(cf.SiacoinOutputs)-1)
-		if err != nil {
-			return
-		}
-		err = sortedUnique(cf.FileContracts, len(cf.FileContracts)-1)
-		if err != nil {
-			return
-		}
-		err = sortedUnique(cf.FileContractTerminations, len(cf.FileContractTerminations)-1)
-		if err != nil {
-			return
-		}
-		err = sortedUnique(cf.StorageProofs, len(cf.StorageProofs)-1)
-		if err != nil {
-			return
-		}
-		err = sortedUnique(cf.SiafundInputs, len(cf.SiafundInputs)-1)
-		if err != nil {
-			return
-		}
-		err = sortedUnique(cf.SiafundOutputs, len(cf.SiafundOutputs)-1)
-		if err != nil {
-			return
-		}
-		err = sortedUnique(cf.ArbitraryData, len(cf.ArbitraryData)-1)
-		if err != nil {
-			return
-		}
-		err = sortedUnique(cf.Signatures, len(cf.Signatures)-1)
-		if err != nil {
-			return
+		for _, field := range fields {
+			if !sortedUnique(field) {
+				return errors.New("field does not satisfy 'sorted and unique' requirement")
+			}
 		}
```

Comment:
you need to call sortedUnique on the siagnatures too, right now you only call it on the other items.

signatures should not be checked in the WholeTransaction loop, but it should be checked in the sortedUnique loop. Maybe append it after the first loop?